BJS2-84998. Создание сервиса по генерации коротких ссылок

FROLOV-KI opened the pull request.

FROLOV-KI committed:
BJS2-84990/84994. Добавлена структура БД для шортнера

FROLOV-KI committed:
BJS2-84989/84991. Добавлен класс для генерации хэша

FROLOV-KI committed:
BJS2-84996. Кэширование свободных хэшей

FROLOV-KI committed:
BJS2-84997/84998. Контроллер и сервис создания ссылок

FROLOV-KI committed:
BJS2-84993. Обработка исключений

FROLOV-KI committed:
BJS2-84988. Очистка устаревших хэшей

psvm-dev commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.List;
+
+public interface HashRepository extends JpaRepository<Hash, String> {
```
не вижу метод save(hashes), сохраняющий список хешей батчами

FROLOV-KI replied:
saveAll уже реализован в JpaRepository

psvm-dev commented on src/main/java/faang/school/urlshortenerservice/service/HashGenerator.java:
```diff
@@ -0,0 +1,30 @@
+        return base62Encoder.encodeBatch(getNextSequenceBatch(count));
+    }
+
+    private List<Long> getNextSequenceBatch(int count) {
```
по условию задачи этот метод должен называться getUniqueNumbers и находиться в HashRepository

FROLOV-KI replied:
не вижу смысла, метод нужен исключительно для генерации новых хэшей, относится к логике этого класса и используется только тут

psvm-dev commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.List;
+
+public interface HashRepository extends JpaRepository<Hash, String> {
```
лучше переделать его на JdbcTemplate т.к. есть требование не мапить данные на какие-либо hibernate сущности

FROLOV-KI replied:
Там написано можно использовать Spring Data, но и JdbcTemplate подойдет. И нам не нужно мапить. Мне кажется требования обычно это как делать нужно, а не как НЕ нужно делать. Для единообразия решил сделать вот так, так как таблица - сущность можно сделать. Использовал jdbcTemplate для сиквенса, чтобы псевдосущность не создавать. Если бы было требование сделать именно через jdbcTemplate согласился бы.

Aleck210popov commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,35 @@
+    public ResponseEntity<ShortUrlResponse> createShortUrl(@Valid @RequestBody ShortUrlRequest shortUrlRequest) {
+        ShortUrlResponse shortUrlResponse = urlControllerFacade.createShortUrl(shortUrlRequest);
+        return ResponseEntity.ok(shortUrlResponse);
+    }
```
Мб 201?

Aleck210popov commented on src/main/java/faang/school/urlshortenerservice/exception/config/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,93 @@
+                HttpMessageNotReadableException.class, HttpStatus.BAD_REQUEST,
+                RecordNotFoundException.class, HttpStatus.NOT_FOUND,
+                PreConditionFailedException.class, HttpStatus.CONFLICT
+        );
```
Если не ошибаюсь Map.of имеет ограничение на 10 entry, при добавлении большего количества исключений, придётся менять. Но это так, особого отношения к задаче не имеет.

Aleck210popov commented on src/main/java/faang/school/urlshortenerservice/exception/common/DataValidationException.java:
```diff
@@ -0,0 +1,7 @@
+public class DataValidationException extends RuntimeException {
+    public DataValidationException(String message) {
+        super(message);
+    }
```
Не нашёл где ты его используешь

Aleck210popov commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,63 @@
+                .map(Hash::new)
+                .toList();
+        hashRepository.saveAll(newHashes);
+        log.info("{} hashes generated", newHashes.size());
```
Работает ли батчем? У меня не получилось настроить

FROLOV-KI replied:
Честно, не проверял, как-то с Димой обсуждали, что вроде как это разовая транзакция в БД

Aleck210popov commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepository.java:
```diff
@@ -0,0 +1,20 @@
+    List<String> getFreeHashBatch(@Param("batchSize") long batchSize);
+
+    void deleteAllByHashIn(List<String> hashes);
+}
```
Эти методы используются в сервисе друг за другом. Но вместе они не атомарны.
Вижу 2 варианта:
1) пессимистик лок
2) delete from select

FROLOV-KI committed:
BJS2-84988. Небольшие правки

FROLOV-KI committed:
BJS2-84988. Обновил HashRepository

1) Добавлена лок при получении свободных хешей
2) Обновлен метод удаление свободныз хэшей из БД
3) Добавил конфиг настройку батчей для saveAll

FROLOV-KI committed:
BJS2-84988. Обновил контроллер

20cognosce commented on src/main/resources/db/changelog/changeset/V001__init.sql:
```diff
@@ -0,0 +1,20 @@
+
+--changeset kfrolov:create_free_hash_storage_table_20250716
+CREATE TABLE free_hash_storage (
+    hash VARCHAR(7) PRIMARY KEY
```
слишком длинный хэш, делай 5 или 6

FROLOV-KI replied:
это просто с запасом сам хеш будет начинаться с 6 символов и очень долго еще будет таким

FROLOV-KI replied:
это просто с запасом в 1 символ, по скиквенсу хэши еще очень долго будут 6 символов, а 7 символов хватит на 100 лет

20cognosce commented on src/main/resources/db/changelog/changeset/V001__init.sql:
```diff
@@ -0,0 +1,20 @@
+
+--changeset kfrolov:create_url_table_20250716
+CREATE TABLE short_url (
+    hash VARCHAR(7) PRIMARY KEY,
+    actual_url TEXT NOT NULL,
+    expiration_time TIMESTAMP NOT NULL,
+    created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
```
форматирование

20cognosce commented on src/main/java/faang/school/urlshortenerservice/util/Base62Encoder.java:
```diff
@@ -0,0 +1,24 @@
+@Component
+public class Base62Encoder {
+
+    private static final String CHAR_SOURCE = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
```
у тебя некорректный алфавит
https://en.wikipedia.org/wiki/Base62

FROLOV-KI replied:
Изначально у меня ткой был, но там первые хеши были только из цифрм, мне каоказалось не красиво и я их в конец перенес

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/config/HashConfig.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+
+@Component
```
@Configuration

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashGenerator.java:
```diff
@@ -0,0 +1,30 @@
+public class HashGenerator {
+
+    private final Base62Encoder base62Encoder;
+    private final JdbcTemplate jdbcTemplate;
```
Зачем ты JDBC подтянул в задачу? в чем необходимость

FROLOV-KI replied:
Чтобы с сикнвенсом работать и получать наборы чисел из БД

20cognosce commented on src/main/java/faang/school/urlshortenerservice/scheduler/shorter/ShorterCleanConfig.java:
```diff
@@ -0,0 +1,26 @@
+@Getter
+@Setter
+@ConfigurationProperties(prefix = "scheduler.expired-urls-clean")
+@Component
```
@Configuration

20cognosce commented on src/main/java/faang/school/urlshortenerservice/scheduler/Scheduler.java:
```diff
@@ -0,0 +1,44 @@
+
+        while (attempt < maxAttempts) {
+            attempt++;
+            try {
+                shorterCleaner.cleanExpiredUrlsBatchAsync(shorterCleanerConfig.getBatchSize());
+            } catch (Exception e) {
```
что ты тут ловишь, если это другой поток уже

20cognosce commented on src/main/java/faang/school/urlshortenerservice/scheduler/shorter/ShorterCleaner.java:
```diff
@@ -0,0 +1,27 @@
+
+    private final UrlService urlService;
+
+    @Async("shorterCleanerExecutor")
```
Зачем тебе Async, если Scheduled итак фоновый поток

20cognosce commented on src/main/java/faang/school/urlshortenerservice/scheduler/Scheduler.java:
```diff
@@ -0,0 +1,44 @@
+        int attempt = 0;
+        int maxAttempts = shorterCleanerConfig.getFetchLimit() / shorterCleanerConfig.getBatchSize();
+
+        while (attempt < maxAttempts) {
+            attempt++;
```
зачем

просто все что есть захватывай в FOR UPDATE SKIP LOCKED и очищай их
если джоба достаточно часто запускается, то можно все захватывать и о батчах никаких не думать

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -9,17 +23,49 @@ spring:
+
+scheduler:
+  expired-urls-clean:
+    # каждый день в 4 часа ночи
+    cron: "0 0 4 * * *"
```
ну это фигня, получается после того как ссылка протухла она еще сутки висит в БД?

FROLOV-KI replied:
3. Джоба запускается раз в сутки по крону.

Проверка то все равно при выдаче есть, отдам что ссылка более не действительна, а ночью снесу из бд

20cognosce commented on src/main/java/faang/school/urlshortenerservice/executor/HashGeneratorExecutorConfig.java:
```diff
@@ -0,0 +1,17 @@
+@Getter
+@Setter
+@ConfigurationProperties(prefix = "executor.hash")
+@Component
```
.

20cognosce commented on src/main/java/faang/school/urlshortenerservice/exception/config/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,93 @@
+        return exception.getMessage();
+    }
+
+    private Map<String, String> extractFieldErrors(Exception exception) {
+        if (exception instanceof MethodArgumentNotValidException ex) {
+            return ex.getBindingResult().getFieldErrors().stream()
+                    .collect(Collectors.toMap(
+                            FieldError::getField,
+                            fieldError -> fieldError.getDefaultMessage() == null
+                                    ? exception.getMessage()
+                                    : fieldError.getDefaultMessage(),
+                            (m1, m2) -> m1 // если дубликаты — берем первый
+                    ));
+        }
+
+        if (exception instanceof ConstraintViolationException ex) {
+            return ex.getConstraintViolations().stream()
+                    .collect(Collectors.toMap(
+                            v -> v.getPropertyPath().toString(),
+                            ConstraintViolation::getMessage,
+                            (m1, m2) -> m1
+                    ));
+        }
```
что за жесть

20cognosce commented on src/main/java/faang/school/urlshortenerservice/exception/config/ErrorResponse.java:
```diff
@@ -0,0 +1,23 @@
+public class ErrorResponse {
+    private int status;
+    private String message;
+    private Map<String, String> fieldError;
```
зачем
сообщения достаточно

20cognosce commented on src/main/java/faang/school/urlshortenerservice/entity/ShortUrl.java:
```diff
@@ -0,0 +1,42 @@
+    private String hash;
+
+    @Column(name = "actual_url", nullable = false, unique = true)
+    private String actualUrl;
```
Здесь @URL навесить тоже имеет смысл, или сразу в URI / URL конвертировать

20cognosce commented on src/main/java/faang/school/urlshortenerservice/dto/ShortUrlResponse.java:
```diff
@@ -0,0 +1,9 @@
+import java.time.LocalDateTime;
+
+public record ShortUrlResponse(
+        String shortUrl,
```
Почему просто строка

20cognosce commented on src/main/java/faang/school/urlshortenerservice/dto/ShortUrlRequest.java:
```diff
@@ -0,0 +1,16 @@
+public record ShortUrlRequest(
+        @NotNull(message = "Required parameter url should be not null")
+        @URL(message = "Invalid url address")
+        String url,
```
.

20cognosce commented on src/main/java/faang/school/urlshortenerservice/dto/RedirectResponse.java:
```diff
@@ -0,0 +1,6 @@
+package faang.school.urlshortenerservice.dto;
+
+public record RedirectResponse(
+        String url
```
.

20cognosce commented on src/main/java/faang/school/urlshortenerservice/controller/UrlControllerFacade.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private ShortUrlResponse createShortUrlResponse(ShortUrl shortUrl) {
+        String shortUrlValue = shorterAddress.concat(shortUrl.getHash());
```
Почему ты с ссылками работаешь как со строками

UriComponentsBuilder

20cognosce commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final UrlControllerFacade urlControllerFacade;
+
+    @GetMapping("/{hash}")
+    public ResponseEntity<RedirectResponse> getResource(@PathVariable String hash) {
+        RedirectResponse redirectResponse = urlControllerFacade.getResource(hash);
+        return ResponseEntity.status(HttpStatus.FOUND).body(redirectResponse);
+    }
```
Из браузера редирект работает?

FROLOV-KI replied:
теперь да)

20cognosce commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepository.java:
```diff
@@ -0,0 +1,26 @@
+    @Query(nativeQuery = true,
+            value = """
+                  SELECT hash FROM free_hash_storage
+                  ORDER BY hash
```
Зачем ты сортируешь таблицу

Ты должен случайные хэши доставать, я так могу предугадать следующий символ и просто протыкать чужие ссылки по порядку алфавита

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,79 @@
+            redisUrlCacheService.cacheUrl(shortUrl);
+        }
+
+        validator.validateNotExpired(shortUrl);
```
такого явления в проверке быть не должно, если ссылка истекла - её просто не должно быть

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,79 @@
+        } else {
+            shortUrl = findByHash(hash);
+            redisUrlCacheService.cacheUrl(shortUrl);
+        }
```
сразу return делать, без переприсвоения

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,79 @@
+                .orElseThrow(() -> new RecordNotFoundException("Invalid short url. Corresponding url not found."));
+    }
+
+    public ShortUrl getResource(String hash) {
```
О каком ресурсе речь
Мы получаем оригинальную ссылку по короткой в этом методе

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,79 @@
+    public void deleteExpiredShortUrls(int limit) {
+        List<String> expiredHashes = urlRepository.findExpiredUrlsHashes(limit);
+        log.info("Batch contains {} hashes.", expiredHashes.size());
+        if(!expiredHashes.isEmpty()) {
```
что с отступами

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    private LocalDateTime setExpirationTimeOrDefault(ShortUrlRequest request) {
+        LocalDateTime maxExpirationTime = LocalDateTime.now().plusYears(DEFAULT_EXPIRATION_TIME_YEAR_DELTA);
+        boolean defaultRequired = request.expirationTime() == null || request.expirationTime().isAfter(maxExpirationTime);
+        return defaultRequired ? maxExpirationTime : request.expirationTime();
+    }
```
ты с фронта что ль принимаешь срок жизни ссылки? wtf

я наспамлю тебе миллиард ссылок с жизнью триллион лет, и что, у тебя хэши просто кончатся на новые ссылки, потому что ты старые никогда не высвободишь

Полное захламление БД

Это дыра, ты должен ограничивать на бэке срок жизни сам

FROLOV-KI replied:
ну так тут и ограничен срок, максимум год, если нужно то можно временную на меньший срок сделать.

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/RedisUrlCacheService.java:
```diff
@@ -0,0 +1,61 @@
+                    redis.call('EXPIRE', KEYS[1], ARGV[1])
+                end
+                return value
+                """;
```
что это

20cognosce replied:
сделай нормальный поиск по ключу (хэшу) в редисе, это велосипеды какие-то

20cognosce replied:
зачем ты переобновляешь ttl таким образом? я минуты 2 пытался понять

1. Разве redis сам не сбросит его при обращении к записи?
2. Если в redis сам не сбрасывает и это нельзя настроить, то делай тогда два запроса пожалуйста

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashCache.java:
```diff
@@ -0,0 +1,75 @@
+    private void startRefillIfNeeded() {
+        boolean needRefilling = hashQueue.size() < hashConfig.getCacheUpdateCount()
+                && refilling.compareAndSet(false, true);
+        if(needRefilling) {
+            refillCacheAsync();
+        }
```
ну а где проверка, что в БД достаточно свободных кэшей

у тебя 2 процесса должны быть:
- следить за очередью
- следить за БД

не хватает в очереди - берешь из БД
не хватает в БД - досоздаешь хэши в рантайме

FROLOV-KI replied:
при запуске проверяется, что в бд хватает на заполнение кэша, если нет то генерируется, в остальное время по идее и бд и хэш начинают заполняться при достижении определенной границы в 20 процентов и по идее там всегда будут хэши в запасе. Нужно чтобы прям очень очень быстро хэши использовались, чтобы и базу и кэш высушить

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Transactional
+    public List<String> getFreeHashes(long count) {
+        List<String> freeHashes = hashRepository.getFreeHashBatch(count);
```
пиши в имени методов, что там блокировка

20cognosce replied:
И здесь такой гарантии нет, если в базе пусто - ты просто отдашь пустой список

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashCache.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public void refillCache() {
+        List<String> freeHashes = hashService.getFreeHashes(hashConfig.getCache().getSize());
+        hashQueue.addAll(freeHashes);
```
ну и где гарантия, что тут тебе не пришел пустой список

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,62 @@
+                refilling.set(false);
+            }
+        });
+    }
```
Тебе не кажется, что это дублирование кода

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,62 @@
+    private void startRefillIfNeeded() {
+        boolean needRefill = hashRepository.count() < hashConfig.getStorageUpdateCount()
+                && refilling.compareAndSet(false, true);
+        if (needRefill) {
```
ну как бы окей, между потоками в одном инстансе это работает, но если 2 инстанса - они оба начнут создавать хэши в БД

в принципе от того что их станет больше чем запланировано это не проблема, но просто нюанс

20cognosce reviewed (changes requested):
В целом похоже на правду, но нужно доделывать

20cognosce reviewed (changes requested):
-

FROLOV-KI committed:
BJS2-84988. Комментарии к ПР

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/UrlService.java:
```diff
@@ -0,0 +1,78 @@
+        return shortUrl;
+    }
+
+    public ShortUrl getShortUrl(ShortUrlRequest request) {
```
create, а не get

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashService.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Transactional
+    public List<String> getFreeHashes(long count) {
+        List<String> freeHashes = hashRepository.getFreeHashBatchWithLockAndDelete(count);
+        boolean lessThanNeeded = freeHashes.size() < hashConfig.getStorage().getSize();
+        if (lessThanNeeded) {
+            int missingCount = hashConfig.getStorage().getSize() - freeHashes.size();
```
Ты не исправил count

20cognosce commented on src/main/java/faang/school/urlshortenerservice/service/HashCache.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private void startRefillIfNeeded() {
+        boolean needRefilling = refilling.compareAndSet(false, true) &&
+                hashQueue.size() < hashConfig.getCacheUpdateCount();
```
.

20cognosce reviewed (approved):
В целом нормально, принято, но небольшие моменты что обсуждали доправь
